pkg/wharfapi/token.go

```diff
@@ -3,148 +3,69 @@ package wharfapi
+	"github.com/iver-wharf/wharf-api/pkg/model/response"
 )
 
-// Token is a value used by the provider plugins to authenticate with the remote
-// providers.
-type Token struct {
-	TokenID  uint   `json:"tokenId"`
-	Token    string `json:"token"`
-	UserName string `json:"userName"`
+type TokenSearch struct {
+	Limit         *int     `url:"limit,omitempty"`
+	Offset        *int     `url:"offset,omitempty"`
+	OrderBy       []string `url:"orderby,omitempty"`
+	UserName      *string  `url:"userName,omitempty"`
+	UserNameMatch *string  `url:"userNameMatch,omitempty"`
 }
 
-// GetTokenByID fetches a token by ID by invoking the HTTP request:
+// GetToken fetches a token by ID by invoking the HTTP request:
 // 	GET /api/token/{tokenID}
-func (c Client) GetTokenByID(tokenID uint) (Token, error) {
-	newToken := Token{}
-
-	apiURL := fmt.Sprintf("%s/api/token/%v", c.APIURL, tokenID)
-	ioBody, err := doRequest("GET | TOKEN |", http.MethodGet, apiURL, []byte{}, c.AuthHeader)
-	if err != nil {
-		return newToken, err
-	}
-
-	defer (*ioBody).Close()
-
-	err = json.NewDecoder(*ioBody).Decode(&newToken)
-	if err != nil {
-		return newToken, err
-	}
-	return newToken, nil
-}
-
-// GetToken tries to search for a token using the username+token pair by
-// invoking the HTTP request:
-// 	POST /api/token/search
-func (c Client) GetToken(token string, userName string) (Token, error) {
-	newToken := Token{}
-
-	path := "/api/tokens/search"
-
-	data := url.Values{}
-	data.Set("Token", token)
-	if userName != "" {
-		data.Add("UserName", userName)
-	}
-
-	u, _ := url.ParseRequestURI(c.APIURL)
-	u.Path = path
-	u.RawQuery = data.Encode()
-
-	ioBody, err := doRequest("GET | TOKEN |", http.MethodPost, fmt.Sprintf("%v", u), []byte{}, c.AuthHeader)
-	if err != nil {
-		return newToken, err
-	}
-
-	defer (*ioBody).Close()
-
-	var tokens []Token
-	err = json.NewDecoder(*ioBody).Decode(&tokens)
-	if err != nil {
-		return newToken, err
-	}
-
-	if len(tokens) == 0 {
-		return newToken, nil
-	}
-
-	return tokens[0], nil
+func (c Client) GetToken(tokenID uint) (response.Token, error) {
+	token := response.Token{}
+	path := fmt.Sprintf("/api/token/%d", tokenID)
+	err := c.GetDecoded(&token, "TOKEN", path, nil)
+	return token, err
 }
 
-// SearchToken tries to match the given token on the token field and username
-// field, by invoking the HTTP request:
-// 	POST /api/tokens/search
-func (c Client) SearchToken(token Token) ([]Token, error) {
-	body, err := json.Marshal(token)
-	if err != nil {
-		return nil, err
-	}
+// GetTokenList filters tokens based on the parameters by invoking the HTTP
+// request:
+// 	GET /api/token
+func (c Client) GetTokenList(params TokenSearch) ([]response.Token, error) {
+	tokens := response.PaginatedTokens{}
 
-	url := fmt.Sprintf("%s/api/tokens/search", c.APIURL)
-	ioBody, err := doRequest("SEARCH | TOKEN |", http.MethodPost, url, body, c.AuthHeader)
+	q, err := query.Values(params)
 	if err != nil {
-		return nil, err
+		return tokens.List, err
```

Comment:
Suggest returning the type as-is instead. There's some metadata that goes lost here.

It's a major version bump, so it's OK.